src/common/pkcs11/pkcs11.cpp

```diff
@@ -0,0 +1,1047 @@
+    Error err = Convert(src.label, dst.mLabel);
+    if (!err.IsNone()) {
+        return err;
+    }
+
+    err = Convert(src.manufacturerID, dst.mManufacturerID);
+    if (!err.IsNone()) {
+        return err;
+    }
+
+    err = Convert(src.model, dst.mModel);
+    if (!err.IsNone()) {
+        return err;
+    }
+
+    err = Convert(src.serialNumber, dst.mSerialNumber);
+    if (!err.IsNone()) {
+        return err;
+    }
+
+    dst.mFlags = src.flags;
+    Convert(src.hardwareVersion, dst.mHardwareVersion);
+    Convert(src.firmwareVersion, dst.mFirmwareVersion);
+    dst.mTotalPublicMemory = src.ulTotalPublicMemory;
+    dst.mFreePublicMemory = src.ulFreePublicMemory;
+    dst.mTotalPrivateMemory = src.ulTotalPrivateMemory;
+    dst.mFreePrivateMemory = src.ulFreePrivateMemory;
+
+    return ErrorEnum::eNone;
+}
+
+Error Convert(const CK_INFO& src, LibInfo& dst)
+{
+    Convert(src.cryptokiVersion, dst.mCryptokiVersion);
+    Convert(src.libraryVersion, dst.mLibraryVersion);
+
+    Error err = Convert(src.manufacturerID, dst.mManufacturerID);
+    if (!err.IsNone()) {
+        return err;
+    }
+
+    err = Convert(src.libraryDescription, dst.mLibraryDescription);
+    if (!err.IsNone()) {
+        return err;
+    }
+
+    return ErrorEnum::eNone;
+}
+
+Error Convert(const Array<ObjectAttribute>& src, Array<CK_ATTRIBUTE>& dst)
+{
+    if (dst.MaxSize() < src.Size()) {
+        return ErrorEnum::eNoMemory;
+    }
+
+    dst.Clear();
+
+    for (const auto& attr : src) {
+        CK_ATTRIBUTE tmp;
+
+        tmp.type = attr.mType;
+        tmp.pValue = (void*)attr.mValue.Get();
+        tmp.ulValueLen = attr.mValue.Size();
+
+        dst.PushBack(tmp);
+    }
+
+    return ErrorEnum::eNone;
+}
+
+Error BuildAttributes(const Array<AttributeType>& types, Array<Array<uint8_t>>& values, Array<CK_ATTRIBUTE>& dst)
+{
+    // number of types and values should be the same in order to be combined into attributes array.
+    if (types.Size() != values.Size()) {
+        return ErrorEnum::eInvalidArgument;
+    }
+
+    if (dst.MaxSize() < types.Size()) {
+        return ErrorEnum::eNoMemory;
+    }
+
+    dst.Clear();
+
+    for (size_t i = 0; i < types.Size(); ++i) {
+        CK_ATTRIBUTE tmp;
+
+        tmp.type = types[i];
+        tmp.pValue = static_cast<void*>(values[i].Get());
+        tmp.ulValueLen = values[i].MaxSize();
+
+        dst.PushBack(tmp);
+    }
+
+    return ErrorEnum::eNone;
+}
+
+Error GetValues(Array<CK_ATTRIBUTE>& src, Array<Array<uint8_t>>& values)
+{
+    if (values.MaxSize() < src.Size()) {
+        return ErrorEnum::eNoMemory;
+    }
+
+    values.Clear();
+
+    for (const auto& attr : src) {
+        if (attr.ulValueLen == CK_UNAVAILABLE_INFORMATION) {
+            return ErrorEnum::eInvalidArgument;
+        }
+
+        auto tmp = Array<uint8_t>(static_cast<uint8_t*>(attr.pValue), attr.ulValueLen);
+        values.PushBack(tmp);
+    }
+
+    return ErrorEnum::eNone;
+}
+
+Array<uint8_t> Convert(const String& val)
```

Comment:
Too many converts also some of them almost same argument type it is hard to guess which one in which case to use. Rename all the convert methods in order to understand conversion details.